### PR TITLE
Add auto-drive package for managing API interaction

### PR DESCRIPTION
**Pull request by clostao**
### **PR Type**
Enhancement, Tests, Documentation

___

### **Description**
- Introduced the `auto-drive` package with API tools for interacting with the auto-drive service, including read, write, and upload operations.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>jest.config.ts</strong><dd><code>Add Jest configuration for ESM and TypeScript support</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/auto-dag-data/jest.config.ts

<li>Added Jest configuration for ESM support.<br> <li> Configured module name mapping and transformation for TypeScript <br>files.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-284ab2d0441d41beb539c92bd9b971b7cfd6d2a828260be3e968376d9ac1844d">+18/-1</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>10 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Implement CID creation and conversion utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/auto-dag-data/src/cid/index.ts

<li>Implemented functions for CID creation and conversion.<br> <li> Utilized Blake3 hashing for CID generation.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-94d995eac83be16cc51bb410bee355201772eef7010efadeadefb8391273cc15">+27/-1</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Add file compression and decompression utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/auto-dag-data/src/compression/index.ts

<li>Added functions for file compression and decompression.<br> <li> Supported zlib algorithm with configurable options.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-da5698788a9ea4ccb679e686edeec82aaddeb879a1fcf4cf9087cc5a4cb8696f">+83/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Implement file encryption and decryption utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/auto-dag-data/src/encryption/index.ts

<li>Implemented file encryption and decryption functions.<br> <li> Supported AES-GCM algorithm with password-based key derivation.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-3bd74ae414b45a85a829e70ce9305906666a36dd6e342ef072f86c100588d7cd">+98/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>chunker.ts</strong><dd><code>Add IPLD chunking and processing utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/auto-dag-data/src/ipld/chunker.ts

<li>Added functions for processing files and metadata into IPLD format.<br> <li> Supported chunking and linking of data for IPLD DAG creation.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-1595a1c6f5b5489fc0d900c81b297595c593de653a5c31503a3adf078c49a6bb">+246/-1</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>OnchainMetadata.ts</strong><dd><code>Add Protobuf definitions and utilities for on-chain metadata</code></dd></summary>
<hr>

packages/auto-dag-data/src/metadata/onchain/protobuf/OnchainMetadata.ts

<li>Added Protobuf definitions for on-chain metadata.<br> <li> Implemented encoding and decoding functions for metadata.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-7ea261c6e5676ddb8b5f492acd3abccb170ff8515f7d6efe665ea9751a89843b">+398/-1</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>read.ts</strong><dd><code>Implement API read calls for auto-drive</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/auto-drive/src/api/calls/read.ts

<li>Implemented API calls for reading data from auto-drive.<br> <li> Included functions for retrieving objects and metadata.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-0f60075717423384b20d1816b81f1602e149c70a13770ea259f6cc900d2cbf49">+194/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>upload.ts</strong><dd><code>Implement API upload calls for auto-drive</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/auto-drive/src/api/calls/upload.ts

<li>Implemented API calls for uploading files and folders.<br> <li> Supported chunked uploads and session completion.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-eefbdaf1888e97d2bf8706f6de365eee5d2c808048af0a6a4224a9ea25323a49">+198/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>connection.ts</strong><dd><code>Add connection setup for auto-drive API</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/auto-drive/src/api/connection.ts

<li>Added connection setup for auto-drive API.<br> <li> Configured request headers and authentication.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-418e8cbe3c6d02230436ccb6490e106fd2cda7f108cad030797af9b02e76d44d">+32/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>objects.ts</strong><dd><code>Define models for auto-drive objects</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/auto-drive/src/api/models/objects.ts

<li>Defined models for object information and summary.<br> <li> Included enums for scope and owner roles.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-fb8929353b90c4caa89e0a6b6c21a5863c1e6497774e1115c1c1d94d5f3c87cd">+47/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>wrappers.ts</strong><dd><code>Add wrapper functions for file and folder uploads</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/auto-drive/src/api/wrappers.ts

<li>Added wrapper functions for file and folder uploads.<br> <li> Supported optional encryption and compression.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-f72af44e464b60e61af931f7c8047d1e3f23f2df1ffaa6a59b0a8746c9f2a591">+164/-0</a>&nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>chunker.spec.ts</strong><dd><code>Add tests for IPLD chunker functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/auto-dag-data/tests/chunker.spec.ts

<li>Added tests for IPLD chunker functionality.<br> <li> Verified file and folder DAG creation processes.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-209d5ceddd91dd24d27403a9d582092b8a817eaef9cfbac5496ab1565a779c49">+295/-1</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>compression.spec.ts</strong><dd><code>Add tests for file compression and decompression</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/auto-dag-data/tests/compression.spec.ts

<li>Added tests for file compression and decompression.<br> <li> Verified zlib algorithm implementation.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-208dbb5b80ad534ed4cb551c7f772c4ac09a5e316fe30a1dee8e781631181c1f">+58/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>encryption.spec.ts</strong><dd><code>Add tests for file encryption and decryption</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/auto-dag-data/tests/encryption.spec.ts

<li>Added tests for file encryption and decryption.<br> <li> Verified AES-GCM algorithm implementation.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-96c0a17bfd7e1d662049549b64e11c2be89dec6f9503f67a52f0de0a1f763cf3">+68/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Add documentation for auto-dag-data package</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/auto-dag-data/README.md

<li>Added documentation for the auto-dag-data package.<br> <li> Included usage examples and installation instructions.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/auto-sdk/pull/145/files#diff-ad6bcff069f61e89c8a449a5910a4d2e213d8ecc7b0a25ebe666847e2ec853e4">+167/-0</a>&nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: packages/auto-drive/src/api/calls/index.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+export * from './download.js'
+export * from './read.js'
```
Comment (marc-aurele-besner) [inline packages/auto-drive/src/api/calls/index.ts:2]: I still don't get why the path are in .js in a .ts file...
I get that ounce we build the package it build .js files, but these files should also be able to be use in native ts file, and so having .js path will brake this...

**Review by marc-aurele-besner — CHANGES_REQUESTED** [review]
I'm not a fan of all these .js file path, I'll like auto-sdk to be fully compatible native TS environment (aka deno 2) and not just when compile to js, and this seems like a step backward

File: packages/auto-drive/src/api/calls/index.ts (hunk 1 above)
Comment (clostao) [reply packages/auto-drive/src/api/calls/index.ts]: Have been updated to `.ts` extension

**Review by marc-aurele-besner — APPROVED** [review]
LGTM 👍

**Pull request merged**